R/data.table.R

```diff
@@ -1861,90 +1861,243 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
+  # type or class)
+  X = x
+  dm = dim(x)
+  cn = names(x)
+  class(X) = NULL
+  p = length(cn)
+  n = dm[1L]
+  
+  # If there is a rownames column, exctract, drop, and update the dimensions
   if (!is.null(rownames)) {
-    # extract that column and drop it.
-    rownames.value = x[[rownames]]
-    dm = dim(x) - 0:1
-    cn = names(x)[-rownames]
-    X = x[, .SD, .SDcols = cn]
-  } else {
-    dm = dim(x)
-    cn = names(x)
-    X = x
-  }
+    X[[rownames.index]] = NULL
+    dm = dm - 0:1
+    cn = cn[-rownames.index]
+    p = p - 1L
+  }
+  
+  # The maximum dimension size (row or column) for a matrix is 2^31-1 (or the Machine maximum integer)
+  # Check and error before doing computation/memory expensive column checks and coercion
+  if (p > .Machine$integer.max) 
+    error("Matrices with > ", .Machine$integer.max, " (.Machine$integer.max) columns are not supported")
+  if (n > INT_MAX) 
+    error("Matrices with > ", .Machine$integer.max, " (.Machine$integer.max) rows are not supported")
+  
+  # Check for any wide matrix like columns and unpack using as.data.table
+  if (length(which_wide_columns(x)) > 0L) {
+    X = as.data.table(X)
+    dm = dim(X)
+    cn = names(X)
+    class(X) = NULL
+    p = length(cn)
+    n = dm[1L]
+  }
+  
+  # If no rows or columns can simply return empty array
   if (any(dm == 0L))
     return(array(NA, dim = dm, dimnames = list(rownames.value, cn)))
-  p = dm[2L]
-  n = dm[1L]
-  collabs = as.list(cn)
-  class(X) = NULL
-  non.numeric = non.atomic = FALSE
-  all.logical = TRUE
-  for (j in seq_len(p)) {
-    if (is.ff(X[[j]])) X[[j]] = X[[j]][]   # nocov to bring the ff into memory, since we need to create a matrix in memory
-    xj = X[[j]]
-    if (length(dj <- dim(xj)) == 2L && dj[2L] > 1L) {
-      if (inherits(xj, "data.table"))
-        xj = X[[j]] = as.matrix(X[[j]])
-      dnj = dimnames(xj)[[2L]]
-      collabs[[j]] = paste(collabs[[j]], if (length(dnj) >
-        0L)
-        dnj
-      else seq_len(dj[2L]), sep = ".")
-    }
-    if (!is.logical(xj))
-      all.logical = FALSE
-    if (length(levels(xj)) > 0L || !(is.numeric(xj) || is.complex(xj) || is.logical(xj)) ||
-        (!is.null(cl <- attr(xj, "class", exact=TRUE)) && any(cl %chin%
-        c("Date", "POSIXct", "POSIXlt"))))
-      non.numeric = TRUE
-    if (!is.atomic(xj))
-      non.atomic = TRUE
-  }
-  if (non.atomic) {
+  
+  # We now need to make sure all the columns are the same type, handling special
+  # cases where some columns are not stored in memory (ff objects), lists, or 
+  # otherwise non-atomic types (e.g. S4). Further, we want to minimize
+  # the number of checks done, e.g. if all columns are already the same type
+
+  col.classes = lapply(X, class)
+  uniq.col.classes = unique(col.classes) # list structure preserves cases where a column has multiple classes
+  uniq.classes = unique(unlist(uniq.col.classes)) # vector of distinct classes used to determine checks to proceed
```

Comment:
would probably benefit to `sort` this as well if we're going to do repeated `%chin%` against it